bugwarrior/services/gerrit.py

```diff
@@ -0,0 +1,123 @@
+
+import json
+import os
+
+import requests
+
+from bugwarrior.config import die
+from bugwarrior.services import IssueService, Issue, ServiceClient
+
+
+class GerritIssue(Issue):
+    SUMMARY = 'gerritsummary'
+    URL = 'gerriturl'
+    FOREIGN_ID = 'gerritid'
+
+    UDAS = {
+        SUMMARY: {
+            'type': 'string',
+            'label': 'Gerrit Summary'
+        },
+        URL: {
+            'type': 'string',
+            'label': 'Gerrit URL',
+        },
+        FOREIGN_ID: {
+            'type': 'numeric',
+            'label': 'Gerrit Change ID'
+        },
+    }
+    UNIQUE_KEY = (URL, )
+
+    def to_taskwarrior(self):
+        return {
+            'project': self.record['project'],
+            'annotations': self.extra['annotations'],
+            self.URL: self.extra['url'],
+
+            'priority': self.origin['default_priority'],
+            'tags': [],
+            self.FOREIGN_ID: self.record['_number'],
+            self.SUMMARY: self.record['subject'],
+        }
+
+    def get_default_description(self):
+        return self.build_default_description(
+            title=self.record['subject'],
+            url=self.get_processed_url(self.extra['url']),
+            number=self.record['_number'],
+            cls='pull_request',
+        )
+
+
+class GerritService(IssueService, ServiceClient):
+    ISSUE_CLASS = GerritIssue
+    CONFIG_PREFIX = 'gerrit'
+
+    def __init__(self, *args, **kw):
+        super(GerritService, self).__init__(*args, **kw)
+        self.url = self.config_get('base_uri').strip('/')
+        self.username = self.config_get('username')
+        self.password = self.config_get_password('password', self.username)
+        self.ssl_ca_path = self.config_get_default('ssl_ca_path', None)
+        self.session = requests.session()
+        self.session.headers.update({
+            'Accept': 'application/json',
+            'Accept-Encoding': 'gzip',
+        })
+
+        self.session.auth = requests.auth.HTTPDigestAuth(
+            self.username, self.password)
+
+        if self.ssl_ca_path != None:
+            self.session.verify = os.path.expanduser(self.ssl_ca_path)
+
+    @classmethod
+    def get_keyring_service(cls, config, section):
+        base_uri = config.get(section, cls._get_key('base_uri'))
+        return "gerrit://%s" % base_uri
+
+    def get_service_metadata(self):
+        return {
+            'url': self.url,
+        }
+
+    @classmethod
+    def validate_config(cls, config, target):
+        for option in ('gerrit.username', 'gerrit.password', 'gerrit.base_uri'):
+            if not config.has_option(target, option):
+                die("[%s] has no '%s'" % (target, option))
+
+        IssueService.validate_config(config, target)
+
+    def issues(self):
+        url = self.url + '/a/changes/' + \
+            '?q=is:open+is:reviewer' + \
+            '&o=MESSAGES&o=DETAILED_ACCOUNTS'
+        response = self.session.get(url)
```

Comment:
Ah, because requests will percent-encode the ``:`` characters, and gerrit doesn't like that.  :(